Initial task runner structure

victorges opened the pull request:
This is the initial commit for the task runner project. It includes:
 - Core structure of the task runner code for subscribing to RabbitMQ events and process tasks accordingly, always publishing their results back to AMQP
 - Initial import task which can handle both downloading from an external URL into an Object Store bucket, or just processing a file that has been directly uploaded to our Object Store.
 - Integration with the API abstractions for all of those

victorges committed:
Create cmd entrypoint boilerplate

Copied from analyzer

victorges committed:
Add CODEOWNERS

victorges committed:
Create main entrypoint for task running

victorges committed:
Just call it task.Runner instead

obvious right

victorges committed:
Create scaffold for starting import task

victorges committed:
Use Livepeer API from import task

All else is just fantasy

victorges committed:
DEV-REVERT: Reference local packages on go.mod

It doesn't compile with the published packages anyway.

victorges committed:
api: Fetch real objects from the API for import task

victorges committed:
task: Implement file upload logic on import task

victorges committed:
task: Implmenet logic for probing file with ffprobe

victorges committed:
task: Use streaming API from OS driver

victorges committed:
task: Return a response from the task

victorges committed:
task: Also parse the message in buildTaskContext

victorges committed:
task: Create unretriable errors instead of bools

Twas very confusing, I'm happy now.

victorges committed:
task: Update task phase to Running as processing begins

victorges committed:
task: Send TaskResultEvent after processing finishes

Among other things as discussed with gioele

victorges committed:
task: Move OSSession creation to runner

victorges committed:
Rename TaskEvent to TaskTriggerEvent

victorges committed:
task: Fix exchange and key of amqp events

victorges committed:
Create Makefile and Dockerfile

victorges committed:
task/import: Process ffprobe output into asset spec

victorges commented on task/import.go:
```diff
@@ -0,0 +1,96 @@
+		metadataFilePath, videoFilePath string
+	)
+	eg.Go(func() error {
+		defer pipe.Close()
```
Got to close the reader with error when errored

victorges commented on task/import.go:
```diff
@@ -0,0 +1,96 @@
+		return nil, err
+	}
+
+	secondaryReader, pipe := io.Pipe()
```
Need a third reader to calculate hash of the file (assuming ffprobe doesn't which I think it doesn't). File size should be already available on probe tho.

Also the hasher doesn't need to be another pipe. Could be just a reader wrapping the original body which accumulates the hash internally and allows querying afterwards.

gioelecerati replied:
This makes me think about the hash of the file in the upload flow, with the import is easy as we got the full file, but on the upload we are proxying directly to the storage. On S3 we got the `Etag` which is infact an md5, so there should be in the runner or the API a service to retrieve it, wdyt? In any case, a sha256sum would be probably harder to get without having the full file?

victorges replied:
Btw, apparently it's not guaranteed that the `Etag` on the S3 API is an MD5. In the case of Filebase, sometimes it will be just an object UUID (when the file is saved with a multipart upload, which is how we are doing now).

victorges commented on task/import.go:
```diff
@@ -0,0 +1,96 @@
+		if err != nil {
+			return fmt.Errorf("error marshaling ffprobe data: %w", err)
+		}
+		metadataFilePath, err = osSess.SaveData(egCtx, metadataFileName, bytes.NewReader(metadata), nil, fileUploadTimeout)
```
This could be outside the eg for simplicity. The upload and probe will terminate together anyway

victorges committed:
task: Add hash of imported files

victorges committed:
task: Move getFile request to a helper func

victorges committed:
Fix boolean logs

victorges committed:
task: Fix the only typo on import task :amaze:

victorges committed:
task: YKW? Add assetSpec to file metadata

Doesn't hurt

victorges committed:
task: Add a minimal non-actually-unit test for import

victorges committed:
go.mod: Use committed version of libs

Just for easy reproducibility

victorges committed:
.github: Add GitHub CI config from livepeer-data

No changes at all just copy seems to work lol

victorges committed:
go.mod: Bring back the replaces cause dev

victorges committed:
task: Implement direct upload handling as an import

victorges committed:
dockerfile: Install static and runtime ffmpegs

victorges committed:
Revert "go.mod: Bring back the replaces cause dev"

Let's keep it only local

gioelecerati reviewed (commented):
Just a few comments and questions

victorges committed:
.github: Install ffmpeg on CI build as well

victorges commented on task/import.go:
```diff
@@ -0,0 +1,131 @@
+}
+
+func getFile(ctx context.Context, osSess drivers.OSSession, params livepeerAPI.TaskParams) (string, io.ReadCloser, error) {
+	if params.DirectUpload != nil {
```
It's not the same task type right now! https://github.com/livepeer/task-runner/blob/c64bba31acaca6d6aeb845ea19175316045b2328/task/runner.go#L98

Tho your comment made me notice that I did create a separate `task.type` and `task.params.directUpload`, but the result is coming on the same `output.Import` field (it is the same result after all).

I think I'm agreeing that we can start simple and just have an `uploadedObjectKey` in the `Import` task params, and just have everything be the same task type. A direct upload becomes an upload followed by an import task, which later we can have a separate optimized implementation anyway even if it is the same task type. WDYT?

victorges commented on task/probe.go:
```diff
@@ -0,0 +1,194 @@
+var (
+	supportedFormats     = []string{"mp4", "mov"}
+	supportedVideoCodecs = map[string]bool{"h264": true}
+	supportedAudioCodecs = map[string]bool{"aac": true}
```
I believe we don't support just copying the audio around right now, but we probably should. Can certainly add those in the future (or just enable them here if we find out we already support).

victorges commented on task/probe.go:
```diff
@@ -0,0 +1,194 @@
+
+func findFormat(supportedFormats []string, format, filename string) (string, error) {
+	actualFormats := strings.Split(format, ",")
+	extension := path.Ext(filename)
```
The file extension comes either from the URL or the `Content-Disposition` header from the response. It has no commitment with being truthful, as I could just rename any text file to be mp4 if I want. The only really trustable format  is the one that comes out of the probe.

The problem is that the output from the probe is not very specific. The format both from `mp4` or `mov` files come out as just `"format_name": "mov,mp4,m4a,3gp,3g2,mj2"` from `ffprobe`. That's why I'm using the extension here, to pick exactly the `extension` format, if it is accurate to what the file actually is. Whole logic is:
 - spit that list of formats from the probe into all the individual formats
 - check if any of those formats are supported + match the file extension. If they do, that's what we're looking for and we return the extension.
 - If the extension doesn't match any actual file format that is suppoted, fallback to the first supported format that shows up in the "actual formats" of the file. In that case, we'll always fallback to `mp4` since it's our top preferred format (first element in the supported formats array)

victorges committed:
task: Make import and directUpload the same task type

victorges committed:
go.mod: Update go-api-client version

victorges committed:
.github: Create a quicker build pipeline for AMD64 image

victorges committed:
.github: Fix build action

victorges committed:
.github: Remove workflow description

victorges committed:
.github: Fix publish amd64 image

victorges committed:
task: Sync with the API schema

victorges committed:
task: Fixes to sync with API

victorges committed:
task/import: Last fixes for functional import task!

 - Get size from the readHasher instead since ffprobe doesn't return a
   size for pipes
 - Fix metadata file full path

victorges committed:
task: Add protection against nack loops

victorges committed:
task: Make it easier to add new task types

Task runner needed some love

victorges committed:
task: Final touches

I'm addicted. Last commit I swear.

victorges committed:
task: Ensure task result events are actually published

Yeah I lied

victorges committed:
task: Declare const with default task handlers

victorges committed:
task: Remove metadata from import task result

hjpotter92 commented on task/import.go:
```diff
@@ -0,0 +1,131 @@
+	if err != nil {
+		return "", nil, fmt.Errorf("error on import request: %w", err)
+	}
+	if resp.StatusCode >= 300 {
```
acc. to the `CheckRedirect` setting in [http.Client](https://pkg.go.dev/net/http#Client), it appears that redirects are followed by default. you can ignore the above comment (however >= 400 status signifies errors)

victorges replied:
This conceptually means an "if not a 2xx". I don't think we need to follow redirects since users should create the import task with the URL that returns the file, not one that will redirect to something else. But I don't think we need to go out of our way to avoid following redirects anyway so we can keep the default behavior. And it seems like that default behavior is to follow up to 10 redirects on a request before just returning the `3xx` response. On that case, I do agree that we should treat it as an error, and also an error we should not retry just like a `4xx`. If the user wants to try again sometime in the future after fixing their server redirection behavior, they can create another task.

victorges commented on task/hash.go:
```diff
@@ -0,0 +1,55 @@
+)
+
+type ReadHasher struct {
+	size   int64
```
done!

victorges committed:
task: Make asset size an uint
